### ISSUE TITLE
+lookup/definition doesn't fall back to dumb-jump

### ISSUE THREAD

**Issue by danlamanna** [hlissner/doom-emacs#4662]
**What did you expect to happen?**
I expected `gd` (`+lookup/definition`) to jump to my function.

**What actually happened?**
I received an error message `LSP :: Not found for: field_file_to_local_path`.

**Additional details:**
- Include a link to your private config
  I ran this inside the sandbox without a private config.

- Include any warnings or errors logged to \*Messages\* (use `M-x
  view-echo-area-messages` to see it).
```
DOOM Looking up ’field_file_to_local_path’ with ’lsp-ui-peek-find-definitions’
DOOM Lookup handler lsp-ui-peek-find-definitions threw an error: (user-error Not found for: field_file_to_local_path)
DOOM Looking up ’field_file_to_local_path’ with ’lsp-find-definition’
LSP :: Not found for: field_file_to_local_path
```

**Steps to reproduce:**
1. Open a python file in an LSP enabled project
2. Try to jump to a function definition with `gd`
   Note that sometimes (often, even) LSP will correctly jump to definition. Keep trying
   until you find a symbol that LSP can't jump to
3. Witness the error message above
4. Confirm this wouldn't be the outcome with `M-x dumb-jump-go`.

**System information:**
<details><pre>
SYSTEM  info       ("Arch Linux" . Linux 5.10.10-arch1-1 x86_64)
        shell      /usr/bin/zsh
        path       (~/.rbenv/shims ~/.local/bin /usr/local/sbin /usr/local/bin /usr/bin /opt/cuda/bin /opt/cuda/integration/nsight-compute /opt/cuda/integration/nsight-systems /var/lib/flatpak/exports/bin /usr/lib/jvm/default/bin /usr/bin/site_perl /usr/bin/vendor_perl /usr/bin/core_perl /usr/lib/emacs/28.0.50/x86_64-pc-linux-gnu)
EMACS   dir        ~/.emacs.d/
        version    (28.0.50 a68d99009604bd38f880086bea49e00008b70f8a Jan 31, 2021)
        buildopts  --prefix=/usr --sysconfdir=/etc --libexecdir=/usr/lib --localstatedir=/var --mandir=/usr/share/man --with-gameuser=:games --with-sound=alsa --with-modules --without-gconf --without-gsettings --with-nativecomp --with-pgtk --with-x-toolkit=gtk3 --without-xaw3d --without-m17n-flt --with-cairo --with-xwidgets --without-compress-install 'CFLAGS=-march=x86-64 -mtune=generic -O2 -pipe -fstack-protector-strong -fno-plt -g -fuse-ld=gold' CPPFLAGS=-D_FORTIFY_SOURCE=2 LDFLAGS=-Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now
        features   ACL CAIRO DBUS FREETYPE GIF GLIB GMP GNUTLS GPM HARFBUZZ JPEG JSON LCMS2 LIBOTF LIBSYSTEMD LIBXML2 MODULES NATIVE_COMP NOTIFY INOTIFY PDUMPER PGTK PNG RSVG SOUND THREADS TIFF TOOLKIT_SCROLL_BARS XIM XWIDGETS GTK3 ZLIB
        traits     (gui server-running)
DOOM    dir        ~/.doom.d/
        version    (2.0.9 HEAD -> develop cf5b7adb6 2021-02-12 23:02:55 -0500)
        font       #<font-object -PfEd-DejaVu Sans Mono-normal-normal-normal-*-13-*-*-*-m-0-iso10646-1>
        theme      doom-one
        modules    (:completion (company +lsp) ivy :checkers syntax (spell +prog) :ui doom doom-dashboard doom-quit hl-todo modeline nav-flash ophints (popup +all +defaults) vc-gutter vi-tilde-fringe window-select workspaces :editor (evil +everywhere) file-templates fold format multiple-cursors rotate-text snippets :emacs (dired +icons) electric (ibuffer +icons) vc :tools ansible debugger direnv docker editorconfig (eval +overlay) (lookup +docsets) (lsp +peek) (magit +forge) prodigy rgb terraform :lang crystal data emacs-lisp (go +lsp) (haskell +lsp) (javascript +lsp) (lua +lsp +moonscript) markdown (org +dragndrop +ipython +pandoc +present) (python +lsp +pyright) rest (rust +lsp) sh (web +lsp) :email notmuch :config (default +bindings +smartparens))
        packages   ((super-save) (dap-mode) (aggressive-indent) (magit-delta) (color-identifiers-mode) (smooth-scrolling) (visual-regexp-steroids) (activity-watch-mode) (edit-server) (string-inflection))
        unpin      (n/a)
        elpa       (spinner s rainbow-identifiers markdown-mode lv lsp-sonarlint lsp-mode ht f dash-functional dash)
</pre></details>

**Comment by hlissner** [comment]
This is because [these LSP handlers](https://github.com/hlissner/doom-emacs/blob/develop/modules/tools/lsp/+lsp.el#L56-L61) don't fail gracefully. Wrapped will need to be written for them. PRs welcome.

**Comment by danlamanna** [comment]
This is a huge pain point for me, so I'd like to help take it on. I looked at it for a little bit and found a related issue https://github.com/jacktasia/dumb-jump/issues/353.

There doesn't seem to be a convenient way of doing this with advice, at least that I can see.

The relevant portion of code from `lsp-mode`:
```elisp
(cl-defun lsp-find-locations (method &optional extra &key display-action references?)
  "Send request named METHOD and get cross references of the symbol under point.
EXTRA is a plist of extra parameters.
REFERENCES? t when METHOD returns references."
  (let ((loc (lsp-request method
                          (append (lsp--text-document-position-params) extra))))
    (if (seq-empty-p loc)
        (lsp--error "Not found for: %s" (or (thing-at-point 'symbol t) ""))
      (lsp-show-xrefs (lsp--locations-to-xref-items loc) display-action references?))))

(cl-defun lsp-find-definition (&key display-action)
  "Find definitions of the symbol under point."
  (interactive)
  (lsp-find-locations "textDocument/definition" nil :display-action display-action))
```

Would it be acceptable, given the simplicity of `lsp-find-definition` to define our own function which effectively rewrites it? Doom would take on the maintenance burden of it going out of sync with upstream, but it's not overtly obvious to me how to achieve the same effect.
```elisp
(cl-defun lsp-find-definition-or-dumb-jump ()
  (let ((loc (lsp-request "textDocument/definition"
                          (lsp--text-document-position-params))))
    (if (seq-empty-p loc)
        (dumb-jump-go)
      (lsp-show-xrefs (lsp--locations-to-xref-items loc) nil nil))))
```

Is something like that along the right lines, or am I way off? It works for me in a very limited set of test cases. If it seems sane I'll open a PR for further feedback.

**Comment by hlissner** [comment]
Sorry for the tremendously late reply. This should be fixed as of 430d628. See [my response to the PR](https://github.com/hlissner/doom-emacs/pull/4739#issuecomment-791824321). Thanks for bringing it to my attention!
